hamp_radar/serde_flight.py

```diff
@@ -38,66 +39,125 @@ def ppar_for_dataset(
+    until a PPAR block is found of the iterable is exhausted.
+
+    Args:
+        blocks (Iterable[DatasetBlockGeometry]): An iterable of dataset blocks to search through.
+
+    Returns:
+        List[DatasetBlockGeometry]: The dataset blocks up to the next PPAR / end of the iterable.
+    """
+    data = []
+    while blocks and blocks.peek().mainblock.subblocks[0].tag != "PPAR":
+        data.append(next(blocks))
+    return data
+
+
 def convert_to_datasetgeometries(
     pfgs: Iterable[PdsFileGeometry],
-) -> List[DatasetGeometry]:
+) -> Iterable[DatasetGeometry]:
     """
-    Converts an iterable of PdsFileGeometry instances into a list of DatasetGeometry instances.
+    Converts an iterable of PdsFileGeometry instances into an iterable of DatasetGeometry instances.
 
     This function converts a collection of PdsFileGeometry instances into
-    DatasetGeometry instances. It identifies "PPAR" tags to start new datasets and appends
-    data blocks to the most recent dataset. If consecutibe PPAR tags are found withou
-    data in between, the ppar of the current dataset is updated to match the last given PPAR.
+    DatasetGeometry instances. It identifies the next "PPAR" tag in the remaining mainblocks from
+    the PdsFileGeometry instances and the associated data thereafter in order to return the
+    geometry of the next dataset from iterating over PdsFileGeometry instances.
 
     Args:
         pfgs (Iterable[PdsFileGeometry]): An iterable of PdsFileGeometry instances, e.g. for a flight.
 
     Returns:
-        List[DatasetGeometry]: A list of DatasetGeometry instances, e.g. for each dataset in a flight.
+        Iterable[DatasetGeometry]: An iterable list of DatasetGeometry instances, e.g. for each dataset in a flight.
     """
-    import warnings
+    from more_itertools import peekable
 
-    datasets = []
-    is_ppar = False
-    for pfg in pfgs:
-        for mmbg in pfg.mainblocks:
-            if mmbg.subblocks[0].tag == "PPAR":
-                ppar = ppar_for_dataset(pfg.filename, mmbg)
-                if is_ppar:
-                    # update ppar of existing dataset
-                    datasets[-1].ppar = ppar
-                else:
-                    # start new dataset
-                    datasets.append(DatasetGeometry(ppar, []))
-                    is_ppar = True
-            else:
-                if datasets == []:
-                    # skip data if no PPAR has yet been found in pfg.mainblocks
-                    warnings.warn(
-                        "Warning: data found before first PPAR is being skipped",
-                        UserWarning,
-                    )
-                else:
-                    # append to latest dataset
-                    is_ppar = False
-                    datablock = DatasetBlockGeometry(pfg.filename, mmbg)
-                    datasets[-1].data.append(datablock)
-    return datasets
+    blocks = peekable(generate_dbgs(pfgs))
```

Comment:
Does this require `more-itertools`? And a new dependency in pyproject?

Reply:
for some reason my environment didn't have more-itertools already so I think yes?? Or maybe my experience was an exceptional case...

Reply:
I think I'm not an exception though, it looks like an additional package despite the name (see https://pypi.org/project/more-itertools/)

Reply:
Oh wait hahaha I realise I added `more-itertools` to the pyproject.toml of another repo instead of this one by accident! :') now I've fixed that whoops!